### ISSUE TITLE
React Module Federation (MFE) supports dynamic module loading like Angular already does

### ISSUE THREAD

**Issue by avington** [nrwl/nx#11160]
## Description
Would like to be able load remote React application via JSON call just like it is supported with the Angular modules. See your [Advanced Angular Modules Frontend with Dynamic Module Federation](https://nx.dev/module-federation/dynamic-module-federation-with-angular#advanced-angular-micro-frontends-with-dynamic-module-federation) article. 

As far as I can tell, this ability is only supported in Angular.

## Motivation
Some our application are deployed on site to a customers data centers and we also have many multi-tenant sites of our own, so we do not know the remote site URL at build time.

Our work around for now is to use URL rewrite rules so that all the remote sites end up in the same domain but this is not optimal.

Also, all of our products use React on the frontend se we would rather not have to switch to Angular (no offense to Angular) for our new stuff.

## Suggested Implementation
Have an equivalent namespace as angular
`import { setRemoteDefinitions } from '@nrwl/react/mf';`

Have similar functions for React as the Angular functions that load the modules at runtime.
`import { setRemoteDefinitions } from '@nrwl/angular/mf';

```
import { setRemoteDefinitions } from '@nrwl/angular/mf // React equivalent import

fetch('/assets/module-federation.manifest.json')
  .then((res) => res.json())
  .then((definitions) => setRemoteDefinitions(definitions)) // <--sets the React modules
  .then(() => import('./bootstrap').catch((err) => console.error(err)));
```
Then lazy loading the module would still work the same as does today.

`const Academics = React.lazy(() => import('ui-academics/Module'));`

**Comment by wangel13** [comment]
Look at my workaround - https://taras.one/blog/dynamic-micro-frontends-with-nx-and-react

**Comment by LongLiveCHIEF** [comment]
@wangel13 I haven't been able to get your workaround to work. Even cloning your example repo and running start just loads the home page, but clicking on any of the remotes returns an error in the overlay:

```
TypeError
Failed to fetch dynamically imported module: http://localhost:4203/remoteEntry.js
Call Stack
 undefined
  localhost:4203/remoteEntry.js:undefined:undefined
```

**Comment by wangel13** [comment]
@LongLiveCHIEF I just cloned my repo and it's working.

Use [parallel start of all apps](https://taras.one/blog/dynamic-micro-frontends-with-nx-and-react#serve-and-check):

```
nx run-many --parallel --target=serve --projects=host,cart,blog,shop --maxParallel=100
```

and go to `localhost:4200`

**Comment by LongLiveCHIEF** [comment]
@wangel13 yep, forgot to serve the other projects.

**Comment by LongLiveCHIEF** [comment]
I've done a lot of digging into source and docs on this, and under the hood, the whole dynamic remotes thing is framework agnostic, as it's the webpack container that provides the functionality.  (see: https://webpack.js.org/concepts/module-federation/#dynamic-remote-containers)

Originally, dynamic remotes were achieved using [webpack external-remotes-plugin](https://www.npmjs.com/package/external-remotes-plugin). Which would allow url resolution of the container at runtime.

Since then, webpack has created functionality to allow [dynamic resolution of remote webpack container urls using a promise](https://webpack.js.org/concepts/module-federation/#promise-based-dynamic-remotes), [or asynchronously](https://webpack.js.org/concepts/module-federation/#dynamic-remote-containers),  as long as the resolved object has `get` and `init` interface described in the docs.

This is what the current angular `withFederatedModules` function does, and is what @wangel13's script does (in addition to ensuring shared scope access). 

From there, content container retrieval is a webpack thing, and the loaded javascript is just built/packed javascript

All this is to say, I'm starting to think that the existing `@nrwl/angular/mf`functions for loading components and setting remote definitions could be put into a package like `@nrwl/mf` and used by either framework.vs

**Comment by wangel13** [comment]
> All this is to say, I'm starting to think that the existing @nrwl/angular/mffunctions for loading components and setting remote definitions could be put into a package like @nrwl/mf and used by either framework.vs

Exactly! Need to to be parameterized - [path for container](https://github.com/nrwl/nx/blob/master/packages/angular/mf/mf.ts#L63) to be framework agnostic.

**Comment by Kavebo** [comment]
In your examples, you are using apps from NX monorepo so remoteEntry.js file has that get/init interface there. But what if I want to have main app with other libraries in NX monorepo but other apps in different repos reated with create-mf-app?
How do I force, or add get/init function to build remoteEntry.js file of some external app so it can be loaded via module-federation.manifest.json and not crash on Error that get or init is not a function?

**Comment by LongLiveCHIEF** [comment]
If you have an mfe in a repo that isn't using nx tooling, you just use normal webpack module federation for that specific mfe to make sure it's webpacked into a container. Remember that at the end of the day, all the nx executor is doing is wrapping webpack.

```js
module.exports = {
  entry: {
    remote: './public-path',
  },
  plugins: [
    new ModuleFederationPlugin({
      name: 'remote', // this name needs to match with the entry name
      exposes: ['./public-path'],
      // ...
    }),
  ],
};
```

You'd still include the remote in the manifest and such. With dynamic remote loading, the `module-federation.config.js` file is only used by the executor to determine which local webpack/http-servers to run, and since you don't need that, you wouldn't include your remote in that file at all.

**Comment by avington** [comment]
I believe with this configuration, your host no longer recognizes it has any remote sites linked to it so --devRemotes won't work. I run each site  separately or use the run-many command then target serve and includes the projects I want to serve. 

Since that run-many command was getting long, I ended setting up a run commands in my host's project.json file to run the projects. 
```
"remotes": {
      "executor": "nx:run-commands",
      "options": {
        "commands": [
          "npx nx serve remote-a",
          "npx nx serve remote-b",
          "npx nx serve remote-c",
          "npx nx serve remote-d"
        ]
      }
    }
```
Then run the command `npx nx run remotes`

BTW: I only really need to run the host and the remote I am working on since the other remotes will never get hit unless I route to them.

**Comment by turnerguo** [comment]
> @nrwl/mf

Hi @LongLiveCHIEF, how about the state of `@nrwl/mf`? I'm using `nx react mfe` with [asynchronously](https://webpack.js.org/concepts/module-federation/#dynamic-remote-containers) loading remotes in host, but it always show me a error `Cannot read properties of undefined (reading 'init')` which initialize the `container` failed:
```js
(async () => {
  // Initializes the shared scope. Fills it with known provided modules from this build and all remotes
  await __webpack_init_sharing__('default');
  const container = window.someContainer; // or get the container somewhere else
  // Initialize the container, it may provide shared modules
  await container.init(__webpack_share_scopes__.default);
  const module = await container.get('./module');
})();
```
I did some research on this, the `@nrwl/react:host` generated host app will not replace the `__webpack_init_sharing__` method at runtime. Meanwhile, the official MFE demo [dynamic-remotes](https://github.com/module-federation/module-federation-examples/tree/master/advanced-api/dynamic-remotes) works correct, and the `__webpack_init_sharing__` method will replaced with `__webpack_require__.I`, the `__webpack_share_scopes__` will be `__webpack_require__.S`, I guess this might be the cause, not sure what's broken inside of `@nrwl/react:host` package?
Hopes some nx member can give a help @AgentEnder @vsavkin please.

**Comment by turnerguo** [comment]
Finally, I found that webpack has some unclear doc on [ModuleFederationPlugin](https://webpack.js.org/plugins/module-federation-plugin/), the `name` must be a string with only `_` connector, and no `-` in the name... which I was using `-` in my remote app, that makes host can not find the container name.
```js
new ModuleFederationPlugin({
      name: 'App2', // you can use only [a-zA-Z_] for a container name
      ...
});
```

**Comment by evandromacedo** [comment]
> Look at my workaround - https://taras.one/blog/dynamic-micro-frontends-with-nx-and-react

@wangel13 dude THANK YOU SO MUCH!!! I've been struggling with this for days and your solution works perfectly here! Thank you! 🎉

**Comment by xiongemi** [comment]
also, reference https://github.com/nrwl/nx/issues/10404 issue for this feature. once this feature is implemented, catch errors for remote might be easier.

for now, there are workarounds mentioned in blogs
- https://medium.com/@wangel13/dynamic-micro-frontends-with-nx-and-react-d04a350732a4
- https://taras.one/blog/dynamic-micro-frontends-with-nx-and-react
